### PR TITLE
8-bit quantization in dnn module and int8 layers

### PR DESCRIPTION
**Pull request by SamFC10**
PR for GSoC'21 project on quantization in DNN module. This PR adds functions to quantize FP32 models and int8 versions of some layers along with tests for the new layers.

<cut/>

| Layer | Status | Remarks |
|:-------:|:--------:|:------------:|
| Convolution | :heavy_check_mark: | Variable weights unsupported |
| Inner Product | :heavy_check_mark: | Variable weights unsupported |
| Pooling | :heavy_check_mark: | Only Max and Average pooling |
| Padding | :heavy_check_mark: |  |
| Flatten | :heavy_check_mark: |  |
| Activations | :heavy_check_mark: |  |
| Concat | :heavy_check_mark: |  |
| Eltwise | :heavy_check_mark: | Eltwise division unsupported |
| BatchNorm, Scale, Shift | :heavy_check_mark: | |
| Data Permutation layers | :heavy_check_mark: |  |

A second PR is planned later this summer to load 8-bit quantized models from other frameworks (ONNX/Tensorflow) and perform inference using int8 layers and weights without converting them to FP32 (as done currently).

mentor : @vpisarev 
relates : #16633 #20188

### Pull Request Readiness Checklist

See details at https://github.com/opencv/opencv/wiki/How_to_contribute#making-a-good-pull-request

- [x] I agree to contribute to the project under Apache 2 License.
- [x] To the best of my knowledge, the proposed patch is not based on a code under GPL or other license that is incompatible with OpenCV
- [x] The PR is proposed to proper branch
- [x] There is reference to original bug report and related work
- [x] There is accuracy test, performance test and test data in opencv_extra repository, if applicable
      Patch to opencv_extra has the same branch name.
- [x] The feature is well documented and sample code can be built with the project CMake

### CODE REVIEW

**Comment by SamFC10** [comment]
@vpisarev MaxPool int8 tests pass now. Tests with maxpooling as the last layer are commented as they compute max indices which is not supported in int8 version.

Although I still don't get why maxpooling as the last layer should compute max index as well by default.
https://github.com/opencv/opencv/blob/4c2dff88de606f2c2337bc7816fbe463a4b9013f/modules/dnn/src/layers/pooling_layer.cpp#L1314

File: modules/dnn/src/layers/batch_norm_layer.cpp [hunk 1]
```diff
@@ -100,15 +100,10 @@ class BatchNormLayerImpl CV_FINAL : public BatchNormLayer
...
         }
-        // We will use blobs to store origin weights and bias to restore them in case of reinitialization.
-        weights_.copyTo(blobs[0].reshape(1, 1));
-        bias_.copyTo(blobs[1].reshape(1, 1));
     }
 
     virtual void finalize(InputArrayOfArrays, OutputArrayOfArrays) CV_OVERRIDE
     {
-        blobs[0].reshape(1, 1).copyTo(weights_);
-        blobs[1].reshape(1, 1).copyTo(bias_);
```
Comment (SamFC10) [inline modules/dnn/src/layers/batch_norm_layer.cpp:111]: This change causes the BatchNorm test to fail. It makes sense as the test was added for this exact change #17284.

In a quantized net, if the int8 version of a layer is not present (like BatchNorm), the fp32 version of the layer is copied along with its params. But here we are overwriting the layer's original parameters with fused ones which causes the layer in the quantized net to give wrong results (NaNs and Infs). Is there a way to solve both the issues?

Comment (vpisarev) [reply modules/dnn/src/layers/batch_norm_layer.cpp:111]: @SamFC10, the compatibility is the first priority, so we need to find some method to provide the information that you need while preserving compatibility. For example, prior to copying blobs to weights_ and bias_ we can save the previous state to some other internal arrays.

**Comment by vpisarev** [comment]
> @vpisarev MaxPool int8 tests pass now. Tests with maxpooling as the last layer are commented as they compute max indices which is not supported in int8 version.
> 
> Although I still don't get why maxpooling as the last layer should compute max index as well by default.
> 
> https://github.com/opencv/opencv/blob/4c2dff88de606f2c2337bc7816fbe463a4b9013f/modules/dnn/src/layers/pooling_layer.cpp#L1314

I actually suggest to support computing indices (in FP32, as before) together with computing the max value in INT8. Yes, it will be slower, but it let us provide 100% compatibility.

File: modules/dnn/src/layers/batch_norm_layer.cpp (hunk 1 above)
Comment (SamFC10) [reply modules/dnn/src/layers/batch_norm_layer.cpp:111]: > prior to copying blobs to weights_ and bias_ we can save the previous state to some other internal arrays

Done. Now batchnorm fusion test pass and quantized nets with fp32 batchnorm layer also work.

**Comment by SamFC10** [comment]
> support computing indices (in FP32, as before) together with computing the max value in INT8

This is not possible right now as a single variable determines the datatypes of all the outputs for a layer. So both outputs[0] and outputs[1] can either be CV_32F or CV_8S. One as CV_32F and another as CV_8S is not possible currently.
https://github.com/opencv/opencv/blob/c2c67c208b430f8d1713a52f3a568f0429c39340/modules/dnn/src/dnn.cpp#L582
https://github.com/opencv/opencv/blob/c2c67c208b430f8d1713a52f3a568f0429c39340/modules/dnn/src/dnn.cpp#L980
Ofcourse changing "dtype" to std::vector would solve it, but that will introduce a lot of complexity in allocating blobs and a lot of work for a feature which is rarely used. Maybe we can keep it as low priority and look at it later.

**Comment by SamFC10** [comment]
Some build issues : 
1. AVX-512 path in int8layers/layers_common.simd.hpp causes build warnings and convolution tests failure (segmentation fault). The same tests are passed locally and in some other builders so I suspect there is an issue with that specific path. Also, I cannot reproduce this locally as my CPU only supports up to AVX2.
2. DNN Tests failure in OpenCL builders. From what I remember, I haven't modified any OpenCL related code, so don't know whats causing the failures.

**Comment by vpisarev** [comment]
> Ofcourse changing "dtype" to std::vector would solve it, but that will introduce a lot of complexity in allocating blobs and a lot of work for a feature which is rarely used. Maybe we can keep it as low priority and look at it later.
ok, sounds good to me. Let's keep it as a low-priority item

**Comment by vpisarev** [comment]
> Some build issues :
> 
> 1. AVX-512 path in int8layers/layers_common.simd.hpp causes build warnings and convolution tests failure (segmentation fault). The same tests are passed locally and in some other builders so I suspect there is an issue with that specific path. Also, I cannot reproduce this locally as my CPU only supports up to AVX2.

I suggest to comment off AVX-512 branches for now (in your newly added code, not everywhere)

> 2. DNN Tests failure in OpenCL builders. From what I remember, I haven't modified any OpenCL related code, so don't know whats causing the failures.

well, you need to figure that out. If you stuck at that, we can look at it together

**Comment by SamFC10** [comment]
@alalek @vpisarev How do I ensure only dnn module tests are run in CI Linux OpenCL builder. I edited my original comment but it looks like tests for all modules is being checked.

**Comment by SamFC10** [comment]
@vpisarev As discussed, int8 layers which had mostly duplicated code (concat, flatten, padding) have been removed and the original fp32 layers are modified to support 8-bit inputs as well.

In some parallel_for(), I have used templates to support multiple datatypes, please check the latest commit to see if any changes have to be made.

**Comment by vpisarev** [comment]
@SamFC10, could you please fix the merge conflicts once again? And then squash commits? We will try to merge your pull request quickly.

**Comment by SamFC10** [comment]
> And then squash commits

Looks like I messed up something.
Command used :
```
git reset --soft HEAD~38
git commit -m ""
git push -f origin int8
```

**Comment by alalek** [comment]
You can rollback changes:
```
git checkout -B int8 79eca09675fecd2b58c237233a0aaaa7197ace6f
```

**Comment by SamFC10** [comment]
Managed to restore my commits and squashed them. Thanks for the help @alalek @vpisarev

**Review by vpisarev — APPROVED** [review]

**Comment by vpisarev** [comment]
:+1:
